iree/compiler/Dialect/HAL/IR/HALOps.cpp

```diff
@@ -961,6 +961,172 @@ void DeviceAllocatorOp::getAsmResultNames(
   setNameFn(result(), "allocator");
 }
 
+//===----------------------------------------------------------------------===//
+// hal.device.switch
+//===----------------------------------------------------------------------===//
+
+void DeviceSwitchOp::build(Builder *builder, OperationState &state,
+                           TypeRange resultTypes, Value device,
+                           ArrayRef<Attribute> conditions,
+                           ArrayRef<ValueRange> conditionArgs,
+                           ArrayRef<NamedAttribute> attributes) {
+  state.addOperands({device});
+  state.addAttribute("conditions", builder->getArrayAttr(conditions));
+  for (auto args : conditionArgs) {
+    state.addOperands(args);
+    auto *region = state.addRegion();
+    auto *entryBlock = OpBuilder(region).createBlock(region);
+    for (auto arg : args) {
+      entryBlock->addArgument(arg.getType());
+    }
+  }
+  state.addTypes(resultTypes);
+  state.addAttributes(attributes);
+  state.resizableOperandList = true;
+}
+
+static ParseResult parseDeviceSwitchOp(OpAsmParser &parser,
+                                       OperationState *result) {
+  OpAsmParser::OperandType device;
+  Type deviceType;
+  if (failed(parser.parseLParen()) || failed(parser.parseOperand(device)) ||
+      failed(parser.parseColonType(deviceType)) ||
+      failed(parser.resolveOperand(device, deviceType, result->operands)) ||
+      failed(parser.parseRParen()) ||
+      failed(parser.parseOptionalArrowTypeList(result->types))) {
+    return failure();
+  }
+
+  result->setOperandListToResizable();
+  SmallVector<Attribute, 4> conditionAttrs;
+  do {
```

Comment:
This block could use a comment (nested do/while loops parsing on commas with some conditions... it takes some thinking to scan through and see what is happening)